hack/licence-detector/validate/validate.go

```diff
@@ -0,0 +1,135 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package validate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/hack/licence-detector/dependency"
+	"golang.org/x/sync/errgroup"
+)
+
+const maxErrors = 12
+
+// Validate runs validation checks against the discovered dependencies.
+func Validate(deps *dependency.List) error {
+	return validateURLs(deps)
+}
+
+func validateURLs(deps *dependency.List) error {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	var errorCount uint32
+	depsChan := make(chan dependency.Info, 64)
+	numWorkers := runtime.NumCPU() + 2
+	client := buildHTTPClient()
+	group, ctx := errgroup.WithContext(ctx)
+
+	// start workers
+	for i := 0; i < numWorkers; i++ {
+		group.Go(func() error {
+			for dep := range depsChan {
+				log.Printf("Checking %s", dep.Name)
+				if err := checkDependencyURL(client, dep); err != nil {
+					log.Printf("ERROR: Invalid URL [%s] for dependency [%s]: %v", dep.URL, dep.Name, err)
+					if count := atomic.AddUint32(&errorCount, 1); count > maxErrors {
+						return errors.New("maximum error count exceeded")
+					}
+				}
+			}
+
+			return nil
+		})
+	}
+
+	// distribute work to workers
+	group.Go(func() error {
+		defer close(depsChan)
+		for _, depList := range [][]dependency.Info{deps.Direct, deps.Indirect} {
+			for _, dep := range depList {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case depsChan <- dep:
+				}
+			}
+		}
+		return nil
+	})
+
+	if err := group.Wait(); err != nil {
+		return fmt.Errorf("validation encountered an error: %w", err)
+	}
+
+	if count := atomic.LoadUint32(&errorCount); count > 0 {
+		return fmt.Errorf("encountered %d validation errors", count)
+	}
+
+	return nil
+}
+
+func buildHTTPClient() *http.Client {
```

Comment:
naming nits: it's not clear to me when you use `build...` vs `mk...` vs `new...`. In other parts `mk...` seems just to initialise a struct or create a new instance but in this file you are actually triggering a side effect e.g. in `mkRequest`

Similarily it's not clear if this particular func could also be called `newHTTPClient`